slidechain/cmd/peg/peg.go

```diff
@@ -127,59 +128,32 @@ func main() {
 	log.Printf("successfully submitted peg-in tx hash %s on ledger %d", succ.Hash, succ.Ledger)
 }
 
-// DoPrepegTx builds, submits the pre-peg TxVM transaction, and waits for it to hit the chain.
-func doPrepegTx(bcid, assetXDR []byte, amount, expMS int64, pubkey ed25519.PublicKey, slidechaind string) error {
-	prepegTx, err := slidechain.BuildPrepegTx(bcid, assetXDR, pubkey, amount, expMS)
-	if err != nil {
-		return errors.Wrap(err, "building pre-peg-in tx")
-	}
-	err = submitPrepegTx(prepegTx, slidechaind)
-	if err != nil {
-		return errors.Wrap(err, "submitting and waiting on pre-peg-in tx")
-	}
-	err = recordPeg(prepegTx.ID, assetXDR, amount, expMS, pubkey, slidechaind)
-	if err != nil {
-		return errors.Wrap(err, "recording peg")
-	}
-	return nil
-}
-
-func submitPrepegTx(tx *bc.Tx, slidechaind string) error {
-	prepegTxBits, err := proto.Marshal(&tx.RawTx)
-	if err != nil {
-		return errors.Wrap(err, "marshaling pre-peg tx")
-	}
-	resp, err := http.Post(slidechaind+"/submit?wait=1", "application/octet-stream", bytes.NewReader(prepegTxBits))
-	if err != nil {
-		return errors.Wrap(err, "submitting to slidechaind")
-	}
-	defer resp.Body.Close()
-	if resp.StatusCode/100 != 2 {
-		return fmt.Errorf("status code %d from POST /submit", resp.StatusCode)
-	}
-	log.Printf("successfully submitted and waited on pre-peg-in tx %x", tx.ID)
-	return nil
-}
-
-func recordPeg(txid bc.Hash, assetXDR []byte, amount, expMS int64, pubkey ed25519.PublicKey, slidechaind string) error {
-	p := slidechain.PegIn{
+// doPrePegIn calls the pre-peg-in Slidechain RPC.
+// That RPC builds, submits, and waits for the pre-peg TxVM transaction and records the peg-in in the database.
+func doPrePegIn(bcid, assetXDR []byte, amount, expMS int64, pubkey ed25519.PublicKey, slidechaind string) ([]byte, error) {
+	p := slidechain.PrePegIn{
+		BcID:        bcid,
 		Amount:      amount,
 		AssetXDR:    assetXDR,
 		RecipPubkey: pubkey,
 		ExpMS:       expMS,
 	}
 	pegBits, err := json.Marshal(&p)
 	if err != nil {
-		return errors.Wrap(err, "marshaling peg")
+		return nil, errors.Wrap(err, "marshaling peg")
 	}
-	resp, err := http.Post(slidechaind+"/record", "application/octet-stream", bytes.NewReader(pegBits))
+	resp, err := http.Post(slidechaind+"/prepegin", "application/octet-stream", bytes.NewReader(pegBits))
 	if err != nil {
-		return errors.Wrap(err, "recording to slidechaind")
+		return nil, errors.Wrap(err, "recording to slidechaind")
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode/100 != 2 {
-		return fmt.Errorf("status code %d from POST /record", resp.StatusCode)
+		return nil, fmt.Errorf("status code %d from POST /prepegin", resp.StatusCode)
+	}
+
+	nonceHashBytes, err := ioutil.ReadAll(resp.Body)
```

Comment:
If instead of this you do:

```
var nonceHash [32]byte
_, err := io.ReadFull(resp.Body, nonceHash[:])
```

and if you then return `nonceHash`, you can simultaneously check that enough bytes were available and eliminate the `copy` in the caller.